crates/bitwarden-crypto/examples/protect_key_with_password.rs

```diff
@@ -0,0 +1,117 @@
+//! This example demonstrates how to securely protect keys with a password using the
+//! [PasswordProtectedKeyEnvelope].
+
+use bitwarden_crypto::{
+    key_ids,
+    safe::{PasswordProtectedKeyEnvelope, PasswordProtectedKeyEnvelopeError},
+    KeyStore, KeyStoreContext,
+};
+
+fn main() {
+    let key_story = KeyStore::<ExampleIds>::default();
+    let mut ctx: KeyStoreContext<'_, ExampleIds> = key_story.context_mut();
+    let mut disk = MockDisk::new();
+
+    // Alice wants to protect a key with a password.
+    // For example to:
+    // - Protect her vault with a pin
+    // - Protect her exported vault with a password
+    // - Protect a send with a URL fragment secret
+    // For this, the `PasswordProtectedKeyEnvelope` is used.
+
+    // Alice has a vault protected with a symmetric key. She wants this protected with a PIN.
+    let vault_key = ctx
+        .generate_symmetric_key(ExampleSymmetricKey::VaultKey)
+        .expect("Generating vault key should work");
+
+    // Seal the key with the PIN
+    // The KDF settings are chosen for you, and do not need to be separately tracked or synced
+    // Next, story this protected key envelope on disk.
+    let pin = "1234";
+    let envelope =
+        PasswordProtectedKeyEnvelope::seal(vault_key, pin, &ctx).expect("Sealing should work");
+    disk.save(
+        "vault_key_envelope",
+        (&envelope).try_into().expect("Saving envelope should work"),
+    );
+
+    // Wipe the context to simulate new session
+    ctx.clear_local();
+
+    // Load the envelope from disk and unseal it with the PIN, and store it in the context.
+    let deserialized: PasswordProtectedKeyEnvelope<ExampleIds> =
+        PasswordProtectedKeyEnvelope::try_from(
+            disk.load("vault_key_envelope")
+                .expect("Loading from disk should work"),
+        )
+        .expect("Deserializing envelope should work");
+    deserialized
+        .unseal(ExampleSymmetricKey::VaultKey, pin, &mut ctx)
+        .expect("Unsealing should work");
+
+    // Alice wants to change her password; also her KDF settings are below the minimums.
+    // Re-sealing will update the password, and KDF settings.
+    let envelope = envelope
+        .reseal(pin, "0000")
+        .expect("The password should be valid");
+    disk.save(
+        "vault_key_envelope",
+        (&envelope).try_into().expect("Saving envelope should work"),
+    );
+
+    // Alice wants to change the protected key. This requires creating a new envelope
+    ctx.generate_symmetric_key(ExampleSymmetricKey::VaultKey)
+        .expect("Generating vault key should work");
+    let envelope = PasswordProtectedKeyEnvelope::seal(ExampleSymmetricKey::VaultKey, "0000", &ctx)
```

Comment:
```suggestion
   let new_key = ctx.generate_symmetric_key(ExampleSymmetricKey::VaultKey)
        .expect("Generating vault key should work");
    let envelope = PasswordProtectedKeyEnvelope::seal(new_key, "0000", &ctx)
        .expect("Sealing should work");
```

Don't we need to pass in the newly generated key?

Reply:
Looking back at this after reviewing the implementation. This is actually the `id` of the key inside the `key_store` that we are passing around. Makes more sense, feel free to disregard.

Reply:
We have a tech debt ticket inherited from Dani's initial work on this here: https://bitwarden.atlassian.net/browse/PM-18102, which would update the api to not require specifying a static keyslot that could have collisions, but always getting a new ID by the API.

That would make this API more intuitive / less error prone to use.